R/bias.R

```diff
@@ -118,121 +112,126 @@ bias_sample <- function(true_values, predictions) {
+  # check that predictions and quantiles have the same length
+  if (!length(predictions) == length(quantiles)) {
+    stop("predictions and quantiles must have the same length")
+  }
+
+  if (anyNA(predictions)) {
+    quantiles <- quantiles[!is.na(predictions)]
+    predictions <- predictions[!is.na(predictions)]
+  }
+
+  if (anyNA(quantiles)) {
+    quantiles <- quantiles[!is.na(quantiles)]
+    predictions <- predictions[!is.na(quantiles)]
+  }
 
-  if (anyNA(upper) || anyNA(lower)) {
-    range <- range[!is.na(upper) & !is.na(lower)]
-    lower_predictions <- lower[!is.na(lower) & !is.na(upper)]
-    upper_predictions <- upper[!is.na(lower) & !is.na(upper)]
+  # if there is no input, return NA
+  if (length(quantiles) == 0 || length(predictions) == 0) {
+    return(NA_real_)
+  }
 
-    # deal with the point forecast case where inputs may be NA
-    if (length(range) == 0 ||
-        length(lower_predictions) == 0 ||
-        length(upper_predictions) == 0
-      ) {
-      return(NA_real_)
-    }
+  # quantiles must be between 0 and 1, increase, and be unique
+  if (any(quantiles < 0) || any(quantiles > 1)) {
+    stop("quantiles must be between 0 and 1")
   }
 
-  # convert range to quantiles
-  lower_quantiles <- abs(100 - range) / (2 * 100)
-  upper_quantiles <- abs(100 + range) / (2 * 100)
+  if (!all(diff(quantiles) > 0)) {
+    stop("quantiles must be increasing")
+  }
 
-  # handling if median is not available - compute median as mean between two
-  # inner most quantiles
-  if (0 %in% range) {
-    median_prediction <- upper_predictions[range == 0]
+  if (0.5 %in% quantiles) {
+    median_prediction <- predictions[quantiles == 0.5]
   } else {
+    # if median is not available, compute as mean of two innermost quantiles
+    message(
+      "Median not available, computing as mean of two innermost quantiles",
+      " in order to compute bias."
+    )
     median_prediction <-
-      0.5 * upper_predictions[which.min(range)] +
-      0.5 * lower_predictions[which.min(range)]
+      0.5 * predictions[quantiles == max(quantiles[quantiles < 0.5])] +
+      0.5 * predictions[quantiles == min(quantiles[quantiles > 0.5])]
   }
 
-
   if (true_value == median_prediction) {
     bias <- 0
     return(bias)
-  } else if (true_value < min(lower_predictions)) {
-    lower <- 0
-    bias <- 1 - 2 * lower
-    return(bias)
-  } else if (true_value > max(upper_predictions)) {
-    upper <- 1
-    bias <- 1 - 2 * upper
-    return(bias)
-  } else if (any(lower_predictions >= true_value)) {
-    lower <- max(lower_quantiles[lower_predictions <= true_value])
-    bias <- 1 - 2 * lower
-    return(bias)
-  } else if (any(upper_predictions <= true_value)) {
-    upper <- min(upper_quantiles[upper_predictions >= true_value])
-    bias <- 1 - 2 * upper
-    return(bias)
+  } else if (true_value < median_prediction) {
+    if (true_value < min(predictions)) {
+      bias <- 1
+    } else {
+      q <- max(quantiles[predictions <= true_value])
```

Comment:
Not really a new change but noting this here (probably better addressed in a new Issue):

This is not the same as in the documentation: "the maximum percentile rank for which the corresponding quantile is still below the true value" which would imply a strict `<`. On first glance think the `<=` is correct but I haven't thought about it for long. 

The maths of this should also be added to the `metric-details` vignette, where we could also note that this is an upper bound to the continuous bias (and approaches it the better [0, 1] is covered by the quantiles).

Reply:
Agree and agree we should address this in a new issue.

Reply:
See #309